PublicMapping/districtbuilder-classic
# Parse config.xml to .env

## Overview

Move project configuration related to architecture (hosts, ports, secrets) into `.env` file. Leave all other project configuration in `config.xml`.

This also ties up a bunch of loose ends for various things that I came across while doing this work. Some examples:
* Reporting is no longer optional or configurable (always use non-BARD reporting)
* Use password for Redis (we were doing this in older iterations of DistrictBuilder, no reason why not to)
* Actually use the key-value store (Redis) configuration for calculator that uses Redis (we weren't doing that before and I don't think that calculator could have been working)
* Make languages dynamic and configurable (previously hard-coded)
* No weak passwords by default (user must provide passwords)
* Removes some settings that were just never used (`email_use_tls` and `email_subject_prefix`)
* Simplifies writing settings (eg. no longer writing `reporting_settings.py` which was used only for BARD)
* Does away with a lot of magic numbers and strings related to configuration

### Notes

#### Backstory

The rationale for doing it the way it was ultimately done changed as work progressed. Initially we were going to put all the configuration in environment variables, but after thinking about it a bit more we realized that we still wanted to leverage the validation in `config.xsd`. Then the thinking was to go all in on `config.xml` and `config.xsd` so as to not have to redo validation or documentation. However, we realized that we needed to have the settings make their way into `docker-compose.yml` and the best (only?) way to do that was via environment variables. At that point it made sense to meet in the middle: use `.env` for settings that are necessary to get the containers running that wouldn't need to change much and continue to use `config.xml` for everything else. As an added bonues, `config.xml` now becomes shareable since it doesn't contain secrets.

While it does seem nice to have everything in `.env` since it would do away with a lot of complexity (`config` container, writing to settings template the result of which is imported by `settings.py`), that `.env` file would get very large, and the more settings that are configured in `.env` the more validation becomes important.

Take languages for example. We would have to have some klugey comma-separated string values like so:

```
LANGUAGES=French,English,Spanish
LANGUAGE_CODES=fr,en,sp
```

This could get unwieldy and would be easy for a user to mess up (eg. order would be important here).

Furthermore, some settings are derived from other non-project settings in `config.xml`, such as `ADJACENCY` and `COMPLEX_CLOROPLETH` so it's unclear how those could even be broken out. 

## Testing Instructions

This PR touches many files and reworks lots of things. Ideally testing would involve exercising as much of the system as possible (i.e. cloning a fresh repo and getting set up from scratch).

At a minimum, you will need to:
* `vagrant reload`
* Restart `vagrant rsync-auto` if running
* `./scripts/update`
* `./scripts/server`

Short of loading shapefiles and nesting geolevels, try running these setup commands which are quicker:
* `./manage.py setup` (tests creating a superuser)
* `./manage.py setup -t`
* `./manage.py setup --views`
* `./manage.py setup -G`

Also, log in to the app and do the following:
* Change language
* Log in as a guest and use the default template
* Log in as a registered user and make a plan
* Verify tiles are being loaded on the map
* Generate a report

Basically make sure everything still works as before.

Also, change settings and make sure they are properly validated and that they show up properly in `config_settings.py` when you run
```
docker-compose build django
```
